### PR TITLE
Public API: ClusterRepo CRD

### PR DESCRIPTION
**Pull request by rohitsakala**
- Migrate ClusterRepo CRD

### CODE REVIEW

**Review by nicholasSUSE — DISMISSED** [review]

**Comment by rohitsakala** [comment]
Waiting for security assessment.

File: pkg/apis/catalog.cattle.io/v1/types.go [hunk 1]
```diff
@@ -6,64 +6,80 @@ import (
...
-	// GitBranch The git branch to follow
+	// GitBranch is the git branch where the helm repository is hosted.
 	GitBranch string `json:"gitBranch,omitempty"`
 
 	// CABundle is a PEM encoded CA bundle which will be used to validate the repo's certificate.
 	// If unspecified, system trust roots will be used.
 	CABundle []byte `json:"caBundle,omitempty"`
 
-	// InsecureSkipTLSverify will use insecure HTTPS to download the repo's index.
+	// InsecureSkipTLSverify will use insecure HTTPS to download the helm repository's index file.
```
Comment (pjbgf) [inline pkg/apis/catalog.cattle.io/v1/types.go:55]:
```suggestion
	// InsecureSkipTLSverify will disable the TLS verification when downloading the helm repository's index file.
	// This is disabled by default. Enabling this is not recommended for production due to the security implications. 
```

File: pkg/apis/catalog.cattle.io/v1/types.go [hunk 2]
```diff
@@ -6,64 +6,80 @@ import (
...
 	InsecureSkipTLSverify bool `json:"insecureSkipTLSVerify,omitempty"`
 
-	// ClientSecretName is the client secret to be used to connect to the repo
-	// It is expected the secret be of type "kubernetes.io/basic-auth" or "kubernetes.io/tls" for Helm repos
-	// and "kubernetes.io/basic-auth" or "kubernetes.io/ssh-auth" for git repos.
-	// For a repo the Namespace file will be ignored
+	// ClientSecret is the client secret to be used to connect to the helm repository
+	// It is expected the secret be of type "kubernetes.io/basic-auth" or "kubernetes.io/tls" for Helm repositories
+	// and "kubernetes.io/basic-auth" or "kubernetes.io/ssh-auth" for Github helm repositories.
```
Comment (pjbgf) [inline pkg/apis/catalog.cattle.io/v1/types.go:60]:
```suggestion
	// ClientSecret is the client secret to be used when connecting to a helm repository.
	// The expected secret type is "kubernetes.io/basic-auth" or "kubernetes.io/tls" for Helm repositories
	// and "kubernetes.io/basic-auth" or "kubernetes.io/ssh-auth" for Github helm repositories.
```

File: pkg/apis/catalog.cattle.io/v1/types.go [hunk 3]
```diff
@@ -6,64 +6,80 @@ import (
...
+	// run the helm install or uninstall commands for a chart.
 	ServiceAccount string `json:"serviceAccount,omitempty"`
 
-	// ServiceAccountNamespace namespace of the service account to use. This value is used only on
-	// ClusterRepo and will be ignored on Repo
+	// ServiceAccountNamespace is the namespace of the service account to use.
 	ServiceAccountNamespace string `json:"serviceAccountNamespace,omitempty"`
 
-	// If disabled the repo clone will not be updated or allowed to be installed from
+	// Usage not known yet. Will be removed after review.
```
Comment (pjbgf) [inline pkg/apis/catalog.cattle.io/v1/types.go:78]: Not sure whether this is a comment to be reviewed pre-merge or we plan to potentially deprecate the field. If the latter, we could keep the previous comment and add an additional DEPRECATED line.

File: pkg/apis/catalog.cattle.io/v1/types.go [hunk 4]
```diff
@@ -6,64 +6,80 @@ import (
...
 	Enabled *bool `json:"enabled,omitempty"`
 
-	// DisableSameOriginCheck attaches the Basic Auth Header to all helm client API calls, regardless of whether the destination of the API call matches the origin of the repository's URL
+	// DisableSameOriginCheck if true attaches the Basic Auth Header to all helm client API calls
+	// regardless of whether the destination of the API call matches the origin of the repository's URL.
```
Comment (pjbgf) [inline pkg/apis/catalog.cattle.io/v1/types.go:82]:
```suggestion
	// regardless of whether the destination of the API call matches the origin of the repository's URL.
	// Defaults to false, which keeps the SameOrigin check enabled. Settings this to true is not recommended
	// in production environments due to the security implications.
```

File: pkg/apis/catalog.cattle.io/v1/types.go [hunk 5]
```diff
@@ -74,26 +90,34 @@ const (
 	FollowerRepoDownloaded RepoCondition = "FollowerDownloaded"
 )
 
+// RepoStatus contains details of the helm repository that is currently being used in the cluster.
 type RepoStatus struct {
+	// ObservedGeneration is used by Rancher controller to track the latest generation of the resource that it triggerd on.
```
Comment (pjbgf) [inline pkg/apis/catalog.cattle.io/v1/types.go:95]:
```suggestion
	// ObservedGeneration is used by Rancher controller to track the latest generation of the resource that it triggered on.
```

Comment (rohitsakala) [reply pkg/apis/catalog.cattle.io/v1/types.go:95]: Done

File: pkg/apis/catalog.cattle.io/v1/types.go (hunk 4 above)
Comment (rohitsakala) [reply pkg/apis/catalog.cattle.io/v1/types.go:82]: Done

File: pkg/apis/catalog.cattle.io/v1/types.go (hunk 3 above)
Comment (rohitsakala) [reply pkg/apis/catalog.cattle.io/v1/types.go:78]: @pjbgf That field is not used. I don't even know if it was used before. I think the field was introduced but it was never implemented. I will keep the old comment. I have another story to look into this.

File: pkg/apis/catalog.cattle.io/v1/types.go (hunk 2 above)
Comment (rohitsakala) [reply pkg/apis/catalog.cattle.io/v1/types.go:60]: Done

File: pkg/apis/catalog.cattle.io/v1/types.go (hunk 1 above)
Comment (rohitsakala) [reply pkg/apis/catalog.cattle.io/v1/types.go:55]: Done

File: pkg/controllers/dashboard/helm/repo.go [hunk 6]
```diff
@@ -293,3 +299,13 @@ func shouldRefresh(spec *catalog.RepoSpec, status *catalog.RepoStatus) bool {
 	refreshTime := time.Now().Add(-interval)
 	return refreshTime.After(status.DownloadTime.Time)
 }
+
+// validateClusterRepo validates few conditions that are meant for ClusterRepo spec.
+func validateClusterRepo(repoSpec catalog.RepoSpec) error {
+	// Both URL and GitRepo should not be filled.
+	if repoSpec.GitRepo != "" && repoSpec.URL != "" {
+		return fmt.Errorf("both GitRepo %s and URL %s are specified. Specify only one", repoSpec.GitRepo, repoSpec.URL)
```
Comment (pjbgf) [inline pkg/controllers/dashboard/helm/repo.go:307]: maybe? 

```suggestion
		return fmt.Errorf("fields GitRepo and URL are mutually exclusive and should not be simultaneously configured")
```

**Pull request closed without merging**